[MRG+2] Refactor docstrings

wmvanvliet opened the pull request:
There is a common pattern we employ where we have a function that does something:

    def plot_object(object):
        """Very long docstring, many parameters"""

and then we have a method that calls this function:

    class object:
        def plot(self):
             "Very long docstring, many parameters"""
             plot_object(self)

To have proper docstrings for both the function and the method, we copy/paste it. This is annoying, hard to read and prone to errors as we update one of the docstrings and not the other (see for example `mne.epochs.plot_drop_log` where the doc erroneously claims it returns a tuple).

This PR introduces two decorators to re-use a docstring:

    @copy_doc
    @copy_function_doc_to_method_doc

The first one simply copies the docstring from one function to another. This is useful when subclassing something and overloading a method.

The second one is a bit more clever and copies the docstring from a function and removes the first parameter from the parameter list. This covers the use case I described above, where we have a method calling a function with `self` as first parameter.

This PR is split into two commits: one adds the code and tests for the decorators, the other refactors the docstrings and contains all the glorious red lines.

agramfort commented:
does the doc still render fine?

does function? still works to get help in ipython shell?

wmvanvliet commented:
yes and yes!

wmvanvliet commented:
it does seem to screw with the doc formatting tests at the moment.

jasmainak commented:
this is a cool addition ! can it be possible to have a test which fails if the copying did not work? I can imagine that this may not work due to a misplaced colon or a typo. It might be annoying to discover this after a PR is merged ...

christianbrodbeck commented:
Does it handle subtle differences between the doc strings? E.g. `pick_types` takes as first argument an `Info` object.

wmvanvliet commented:
@jasmainak I've added some error messages to the decorators in the event the copying of the docstring failed.

To further test the new docstrings, I've written a little script:
https://gist.github.com/wmvanvliet/de8d322773cd79848e486d56d6590835

This script extracts all docstrings of the MNE module and saves them to a text file. It then goes on to compute distance metrics between them to produce a list of docstrings that are very similar and possible candidates for the `@copy_doc` and `@copy_function_doc_to_method_doc` decorators.

Using the script to extract the docstrings for the master branch and this PR branch and `diff`-ing them, I checked that all differences were simply due to inconsistencies on the copy/pasted docstrings.

There were also some inconsistencies in the function signatures, like parameters that the function version took and the method version never exposed. Or sometimes the order of the parameters would differ. I've added a commit to straighten these things out.

wmvanvliet commented:
@christianbrodbeck It does not do anything too clever: it just removes the first parameter. In the case of `pick_types` this works out great.

Eric89GXL commented:
So much red :)_ (had to look up the "drooling" emoticon)

jasmainak commented on mne/utils.py:
```diff
@@ -790,6 +790,137 @@ def dec(*args, **kwargs):
+        # Find parameter block
+        for line, text in enumerate(doc[:-2]):
+            if (text.strip() == 'Parameters' and
+                    doc[line + 1].strip() == '----------'):
```
what if there is a typo here or the number of hyphens is somehow less? This will pass silently and you will incorrectly assume that there is no parameter block

Eric89GXL replied:
We should improve our docstring checker, it should be pretty easy for it to catch this sort of case

wmvanvliet replied:
Since the last commit, this will now fail loudly. If there is no parameter block, big error message.

jasmainak replied:
awesome!

jasmainak commented on mne/channels/channels.py:
```diff
@@ -439,13 +439,13 @@ class UpdateChannelsMixin(object):
-                   seeg=False, bio=False, ecog=False, include=[],
-                   exclude='bads', selection=None):
+                   seeg=False, dipole=False, gof=False, bio=False, ecog=False,
+                   include=[], exclude='bads', selection=None):
```
this is a bugfix? But it will break scripts, no?

Eric89GXL replied:
Strictly speaking it's possible, but realistically this is so far down the parameter list I think the risk of breakage is low enough to be okay (presumably/hopefully people just use kwargs here). We do things like this with plotting functions from time to time.

wmvanvliet replied:
I saw it as a bug. There are more plotting functions for which I've messed with the parameter list to make them consistent. A good example of this is `mne.viz.plot_ica_properties` where I flipped the first two parameters to make it consistent. For that one I figured since the function isn't part of any stable release yet I could get away with it :)

jasmainak replied:
okay I agree. But if you do it all together for 8 functions, you are raising the risk quite a bit.

Eric89GXL replied:
Yes but `p = 8 * 1e-12` is still small :)

Eric89GXL replied:
Oh wait that's not how probability works... I mean something more like `1-(1-1e-12) ** 8` is still small. Now if the original probability is more like `1e-1` we're in more trouble, but I don't think it's so high.

jasmainak replied:
lol. We can raise an error "uh oh, we didn't deprecate because of botched up probability calculation but you're welcome to help us correct the calculation" :P I guess the number `1e-12` depends on the number of users we have ...

wmvanvliet replied:
plus, if I have to copy/paste back in 500+ lines of comments just because the order of the parameters is unintentionally inconsistent between two functions... my programmers brain wants to get slam drunk. I have little kids. **THINK OF THE CHILDREN!**

wmvanvliet commented:
Looks like the 4 horsemen of CI are satisfied.

Eric89GXL commented on mne/channels/channels.py:
```diff
@@ -420,43 +420,14 @@ def plot_sensors(self, kind='topomap', ch_type=None, title=None,
-        :func:`mne.viz.plot_trans`.
-
-        .. versionadded:: 0.12.0
-
```
Why get rid of versionadded, notes, and see also here? It would be good to make your function smart enough to handle them (just copy?) if possible.

Eric89GXL commented on mne/cov.py:
```diff
@@ -1960,38 +1935,9 @@ def _estimate_rank_meeg_signals(data, info, scalings, tol='auto',
-    s : array
-        If return_singular is True, the singular values that were
-        thresholded to determine the rank are also returned.
-    """
```
Hmm... this one actually makes the code less readable/usable *during development*, i.e. when staring at code. I guess it's okay in this instance, but there are probably some cases where the evils of duplication would be worth it, e.g. in the complicated bits of `forward` code if we eventually have examples of duplication there.

wmvanvliet replied:
Yes. You should only copy the doc when all your function does is call another function. Or the case you've highlighted here: also copied almost the entire function body.

Eric89GXL commented on mne/utils.py:
```diff
@@ -790,6 +790,141 @@ def dec(*args, **kwargs):
+            a : int
+                Some parameter
+            b : int
+                Some parameter
```
This should go in an examples or notes section, not the topmost one.

Eric89GXL commented on mne/io/base.py:
```diff
@@ -1456,228 +1457,41 @@ def save(self, fname, picks=None, tmin=0, tmax=None, buffer_size_sec=10,
                         lowpass, filtorder, clipping)
 
     @verbose
+    @copy_function_doc_to_method_doc(plot_raw_psd)
```
Does the order of these two decorators matter? Hopefully not, but if it does, please document it.

wmvanvliet replied:
no, it doesn't matter

Eric89GXL commented:
Otherwise LGTM and at least the one function I check (`anonymize`) looks good. This is awesome @wmvanvliet :)

jona-sassenhagen commented:
>So much red :)_ (had to look up the "drooling" emoticon)

Haha @Eric89GXL

wmvanvliet commented:
I've changed the behavior a little bit, now the copy decorator will prepend the docstring of the source function, instead of overriding the docstring. This allows us to keep the `versionadded` bits.

Eric89GXL commented on mne/channels/channels.py:
```diff
@@ -434,29 +434,15 @@ def plot_sensors(self, kind='topomap', ch_type=None, title=None,
-
-        .. versionadded:: 0.13.0
-        """
+        """.. versionadded:: 0.13.0"""
```
I think this probably still needs to go in `Notes` section, it doesn't render properly:

https://2095-1301584-gh.circle-artifacts.com/0//home/ubuntu/mne-python/doc/_build/html/generated/mne.io.Raw.html#mne.io.Raw.anonymize

Eric89GXL replied:
Scratch that, looks like there is another you can get rid of:

https://github.com/wmvanvliet/mne-python/blob/b89390b3033b8c6e4f71f2fcacb76fcae6773db6/mne/io/base.py

wmvanvliet replied:
which one?

wmvanvliet replied:
ah, found it.

agramfort replied:
to move to Notes according to last commit

Eric89GXL replied:
Yep, still does not render properly, needs more newlines:

https://2153-1301584-gh.circle-artifacts.com/0//home/ubuntu/mne-python/doc/_build/html/generated/mne.io.Raw.html#mne.io.Raw.anonymize

Eric89GXL commented on mne/cov.py:
```diff
@@ -31,8 +31,10 @@
 from .utils import (check_fname, logger, verbose, estimate_rank,
-                    _compute_row_norms, check_version, _time_mask, warn)
+                    _compute_row_norms, check_version, _time_mask, warn,
+                    copy_doc, copy_function_doc_to_method_doc)
```
unnecessary import:

https://travis-ci.org/mne-tools/mne-python/jobs/150015614#L3033

jasmainak commented:
@wmvanvliet it occurred to me that this can be useful also in deprecated functions. Often, the only thing that changes is the function signature and you have the same parameters. So, for one release you have duplicated docstrings which can be avoided ...

wmvanvliet commented:
(rebased)

wmvanvliet commented:
@jasmainak 'yup, although I didn't find any such cases.

all comments are addressed.

agramfort commented on mne/io/base.py:
```diff
@@ -710,20 +711,6 @@ def __setitem__(self, item, value):
         # set the data
         self._data[sel, start:stop] = value
 
-    def anonymize(self):
```
why this?

wmvanvliet replied:
the function is already defined in the Mixin, no need to re-implement it here.

agramfort commented on mne/io/fiff/raw.py:
```diff
@@ -432,7 +433,7 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
                     break
 
     def fix_mag_coil_types(self):
-        """Fix Elekta magnetometer coil types
```
why?

wmvanvliet replied:
oops, that's a leftover from my shuffling with docstrings. Reverting this change.

Eric89GXL commented on mne/utils.py:
```diff
@@ -792,6 +792,167 @@ def dec(*args, **kwargs):
+    >>> class A:
+    ...     @copy_function_doc_to_method_doc(plot_function)
+    ...     def plot(self, a, b):
+    ...         '''.. versionadded:: 0.13.0'''
```
This should actually create a `Notes` section right?

Eric89GXL commented on mne/utils.py:
```diff
@@ -792,6 +792,167 @@ def dec(*args, **kwargs):
+    --------
+    >>> def plot_function(object, a, b):
+    ...     '''Docstring for plotting function.
+    ...     Parameters
```
And you should put another newline before `Parameters`

Eric89GXL commented on mne/utils.py:
```diff
@@ -792,6 +792,167 @@ def dec(*args, **kwargs):
+    Notes
+    -----
+    The parsing performed is very basic and will break easily on docstrings
+    that are not formatted exactly according to the basic MNE structure. Always
```
... according to the `numpydoc` standard

Eric89GXL commented:
LGTM, +1 for merge even without my nitpick changes

Eric89GXL commented:
(it would be good to get this in ASAP since it's so useful and also touches so many files)

agramfort commented on mne/channels/channels.py:
```diff
@@ -379,7 +379,7 @@ def plot_sensors(self, kind='topomap', ch_type=None, title=None,
             Whether to plot the sensors as 3d, topomap or as an interactive
             sensor selection dialog. Available options 'topomap', '3d',
-            'select'. If 'select', a set of channels can be selected
+            'select'.  If 'select', a set of channels can be selected
```
revert extra space

agramfort commented:
besides +1 for merge.

tell me when done + CIs are happy. Thx @wmvanvliet

Eric89GXL commented:
Thanks @wmvanvliet
